File: web/src/algorithms/snippets/radix.js
/** LSD radix sort, base 10 (visualization-friendly). */
export function radixSort(data) {
  const n = data.length;
  if (n < 2) return data.slice();
  let a = data.slice();
  let b = new Array(n);
  const max = Math.max(...a);
  const passes = Math.max(1, Math.ceil(Math.log10(max + 1)));

  for (let pass = 0; pass < passes; pass++) {
    const count = Array(10).fill(0);
    for (let i = 0; i < n; i++) count[digit(a[i], pass)]++;
    const off = Array(10).fill(0);
    let sum = 0;
    for (let d = 0; d < 10; d++) {
      off[d] = sum;
      sum += count[d];
    }
    const run = off.slice();
    for (let i = 0; i < n; i++) {
      const d = digit(a[i], pass);
      b[run[d]++] = a[i];
    }
    [a, b] = [b, a];
    b = new Array(n);
  }
  return a;
}

function digit(v, pass) {
  return Math.floor(v / 10 ** pass) % 10;
}

File: web/src/algorithms/snippets/radix.c
/* LSD radix sort — base 256, 4 passes (32-bit keys) */
#include <stdint.h>
#include <string.h>

void radix_sort(int *data, size_t n) {
    if (n < 2) return;
    uint32_t *a = malloc(n * sizeof(uint32_t));
    uint32_t *b = malloc(n * sizeof(uint32_t));
    for (size_t i = 0; i < n; ++i)
        a[i] = (uint32_t)data[i] ^ 0x80000000u;
    for (int pass = 0; pass < 4; ++pass) {
        int shift = pass * 8;
        size_t count[256] = {0};
        for (size_t i = 0; i < n; ++i)
            count[(a[i] >> shift) & 0xFF]++;
        size_t off[256], sum = 0;
        for (int d = 0; d < 256; ++d) {
            off[d] = sum;
            sum += count[d];
        }
        for (size_t i = 0; i < n; ++i) {
            uint32_t d = (a[i] >> shift) & 0xFF;
            b[off[d]++] = a[i];
        }
        uint32_t *t = a;
        a = b;
        b = t;
    }
    for (size_t i = 0; i < n; ++i)
        data[i] = (int)(a[i] ^ 0x80000000u);
    free(a);
    free(b);
}
